### PR TITLE
Apr 2826 recall companies house people in control data

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs [hunk 1]
```diff
@@ -74,6 +84,60 @@ public async Task<IActionResult> StartPage(Guid applicationId)
...
+                    organisation.OrganisationDetails.CompaniesHouseDetails.Directors = companyDetails.Directors;
+                    organisation.OrganisationDetails.CompaniesHouseDetails.PersonsSignificationControl = companyDetails.PersonsWithSignificantControl;
+
+                    await _organisationApiClient.Update(organisation, User.GetUserId());
+
+                    var directorsAnswers = RoatpPreambleQuestionBuilder.CreateCompaniesHouseWhosInControlQuestions(applicationDetails);
+                    await _qnaApiClient.UpdatePageAnswers(applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CompaniesHouseStartPage, directorsAnswers.ToList<Answer>());
+                    
+                    await _qnaApiClient.ResetPageAnswersBySequenceAndSectionNumber(applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CompaniesHouseStartPage);
+                    // TODO: reset section 3.4
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:129]: don't forget section 3.4 !!

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs [hunk 2]
```diff
@@ -74,6 +84,60 @@ public async Task<IActionResult> StartPage(Guid applicationId)
             return await AddPeopleInControl(applicationId);
         }
 
+        [HttpGet("refresh-directors-pscs")]
+        public async Task<IActionResult> RefreshDirectorsPscs(Guid applicationId)
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:88]: You might be able to grab UKPRN from the model in the view

Comment (martin-lush) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:88]: actually you'd need to grab this in the StartPage(Guid applicationId) method (we get qna application data here so it'll be in the question tags) and then through to the relevant methods & viewmodels

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs (hunk 1 above)
Comment (martin-lush) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:126]: To be done in different ticket

File: src/SFA.DAS.ApplyService.Web.UnitTests/Controllers/RoatpWhosInControlApplicationControllerTests.cs [hunk 3]
```diff
@@ -1619,5 +1659,104 @@ public void Remove_item_from_pscs_saves_new_answer_with_entry_removed_if_confirm
...
+                Directors = listOfDirectors,
+                PersonsWithSignificantControl = listOfPSCs,
+                IncorporationDate = new DateTime(1960, 12, 12),
+                Status = "active"
+            };
+
+            _applicationClient.Setup(x => x.GetOrganisationByUkprn(ukprn)).ReturnsAsync(organisation);
+            _organisationApiClient.Setup(x => x.Update(It.IsAny<Organisation>(),It.IsAny<Guid>())).ReturnsAsync(new Organisation());
+            _companiesHouseApiClient.Setup(x => x.GetCompanyDetails(companyNumber))
+                .Returns(Task.FromResult(activeCompany)).Verifiable();
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web.UnitTests/Controllers/RoatpWhosInControlApplicationControllerTests.cs:1716]: Could you make it as _**.ReturnsAsync(activeCompany)).Verifiable();**_ ?

File: src/SFA.DAS.ApplyService.Web.UnitTests/Controllers/RoatpWhosInControlApplicationControllerTests.cs [hunk 4]
```diff
@@ -1619,5 +1659,104 @@ public void Remove_item_from_pscs_saves_new_answer_with_entry_removed_if_confirm
...
+            var activeCompany = new CompaniesHouseSummary
+            {
+                CompanyNumber = companyNumber,
+                CompanyType = "ltd",
+                IncorporationDate = new DateTime(1960, 12, 12),
+                Status = companiesHouseStatus
+            };
+      
+            _companiesHouseApiClient.Setup(x => x.GetCompanyDetails(companyNumber))
+                .Returns(Task.FromResult(activeCompany)).Verifiable();
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web.UnitTests/Controllers/RoatpWhosInControlApplicationControllerTests.cs:1750]: Could you make it as **_.ReturnsAsync(activeCompany)).Verifiable();_** ?

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs [hunk 5]
```diff
@@ -74,8 +84,59 @@ public async Task<IActionResult> StartPage(Guid applicationId)
             return await AddPeopleInControl(applicationId);
         }
 
+        [HttpGet("refresh-directors-pscs")]
+        public async Task<IActionResult> RefreshDirectorsPscs(Guid applicationId, string ukprn, string companyNumber)
+        {
+            try
+            {
+                var companyDetails = await _companiesHouseApiClient.GetCompanyDetails(companyNumber);
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:92]: Comment - Does this api call handle null ok?

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs [hunk 6]
```diff
@@ -74,8 +84,59 @@ public async Task<IActionResult> StartPage(Guid applicationId)
...
+
+                    if (!CompaniesHouseValidator.CompaniesHouseStatusValid(companyDetails.CompanyNumber, companyDetails.Status))
+                    {
+                        _logger.LogInformation($"Issue refreshing directors/pscs - applicationId {applicationId} | Company Number : {companyDetails.CompanyNumber} | Status : Companies House status not valid: {companyDetails.Status}");
+                        return RedirectToAction("CompanyNotFound", "RoatpShutterPages");
+                    }
+
+                    var applicationDetails = new Domain.Roatp.ApplicationDetails { CompanySummary = companyDetails };
+
+                    var organisation = await _applicationApiClient.GetOrganisationByUkprn(ukprn);
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:115]: Comment - Does this api call handle null ok?

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs [hunk 7]
```diff
@@ -74,8 +84,59 @@ public async Task<IActionResult> StartPage(Guid applicationId)
...
+                _logger.LogError(ex, $"Error when processing directors/pscs - applicationId {applicationId}");
+                return RedirectToAction("CompaniesHouseNotAvailable", "RoatpShutterPages");
+            }
+
+            return RedirectToAction("ConfirmDirectorsPscs", "RoatpWhosInControlApplication", new { applicationId });
+        }
+
         [Route("confirm-directors-pscs")]
-        public async Task<IActionResult> ConfirmDirectorsPscs(Guid applicationId)
+        public async Task<IActionResult> ConfirmDirectorsPscs(Guid applicationId, string ukprn, string companyNumber)
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:139]: line 138
I think this ought to be: **_[HttpPost("confirm-directors-pscs")]_**

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs (hunk 5 above)
Comment (martin-lush) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:92]: discussed that this pattern was lifted from elsewhere so better to keep with the existing behaviour for now...

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs (hunk 6 above)
Comment (martin-lush) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:115]: discussed that this pattern was lifted from elsewhere so better to keep with the existing behaviour for now...

**Review by martin-lush — DISMISSED** [review]

**Review by martin-lush — APPROVED** [review]

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:115]: yes, but that's adding performant endpoints over the usual crud-like wow in roatp.  I'll do it, but want to have a bigger discussion

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs [hunk 8]
```diff
@@ -74,8 +84,59 @@ public async Task<IActionResult> StartPage(Guid applicationId)
...
+
+                    organisation.OrganisationDetails.CompaniesHouseDetails.Directors = companyDetails.Directors;
+                    organisation.OrganisationDetails.CompaniesHouseDetails.PersonsSignificationControl = companyDetails.PersonsWithSignificantControl;
+
+                    await _organisationApiClient.Update(organisation, User.GetUserId());
+
+                    var directorsAnswers = RoatpPreambleQuestionBuilder.CreateCompaniesHouseWhosInControlQuestions(applicationDetails);
+                    await _qnaApiClient.UpdatePageAnswers(applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CompaniesHouseStartPage, directorsAnswers.ToList<Answer>());
+                    
+                    await _qnaApiClient.ResetPageAnswersBySequenceAndSectionNumber(applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CompaniesHouseStartPage);
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs]: I believe there is a deliberate reason this is ordered this way, but will look into it.

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs [hunk 9]
```diff
@@ -74,8 +84,59 @@ public async Task<IActionResult> StartPage(Guid applicationId)
             return await AddPeopleInControl(applicationId);
         }
 
+        [HttpGet("refresh-directors-pscs")]
+        public async Task<IActionResult> RefreshDirectorsPscs(Guid applicationId, string ukprn, string companyNumber)
+        {
+            try
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:90]: ok

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs [hunk 10]
```diff
@@ -74,8 +84,59 @@ public async Task<IActionResult> StartPage(Guid applicationId)
...
+            {
+                var companyDetails = await _companiesHouseApiClient.GetCompanyDetails(companyNumber);
+
+                if (companyDetails != null)
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:94]: ok

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (DevenKShah) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs:115]: Indeed we can discuss this. But patch is sort of a CRUD end point.

File: src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Web/Controllers/Roatp/RoatpWhosInControlApplicationController.cs]: It appears line 125 needs to run after line 123, so I won't touch.

**Review by DevenKShah — APPROVED** [review]

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**